# EDA 2

Primero se realiza la importacion de las librerias a utilizar

In [30]:

import pandas as pd
import seaborn as sbn
import numpy as np
import matplotlib.pyplot as plt
from seaborn import load_dataset
import io

Llamado de data por google colab

In [31]:

#from google.colab import files
#uploaded = files.upload()
#df = pd.read_csv(io.BytesIO(uploaded['bl2.csv']))

Llamado por local 

In [32]:
df = pd.read_csv('Datas/bl2.csv')

Visualisamos los primeros campos de la data

In [33]:
df.head()

Unnamed: 0              name           full_name  age  height  nationality  \
0           0      Manuel Neuer  Manuel Peter Neuer   37    1.93      Germany   
1           1       Yann Sommer                 NaN   34    1.83  Switzerland   
2           2      Sven Ulreich                 NaN   34    1.92      Germany   
3           3   Johannes Schenk                 NaN   20    1.91      Germany   
4           4  Matthijs de Ligt                 NaN   23    1.89  Netherlands   

  place_of_birth  price  max_price                position  shirt_nr   foot  \
0  Gelsenkirchen    7.0       45.0              Goalkeeper         1  right   
1         Morges    5.0       13.0              Goalkeeper        27  right   
2     Schorndorf    0.9        6.0              Goalkeeper        26  right   
3    Schweinfurt    0.3        0.3              Goalkeeper        35    NaN   
4     Leiderdorp   75.0       75.0  Defender - Centre-Back         4  right   

            club contract_expires joined_club             player_agent  \
0  Bayern Munich       2024-06-30  2011-07-01          PRO Profil GmbH   
1  Bayern Munich       2025-06-30  2023-01-19                Relatives   
2  Bayern Munich       2024-06-30  2021-07-01  BMS Sportconsulting ...   
3  Bayern Munich       2024-06-30  2022-07-01                   11WINS   
4  Bayern Munich       2027-06-30  2022-07-19          Rafaela Pimenta   

  outfitter  
0    adidas  
1      Puma  
2    adidas  
3       NaN  
4    adidas

y ahora los ultimos

In [34]:
df.tail()

Unnamed: 0                name                  full_name  age  height  \
510         510        Takuma Asano                      浅野 拓磨   28    1.73   
511         511     Philipp Hofmann                        NaN   30    1.95   
512         512        Simon Zoller        Simon Martin Zoller   31    1.79   
513         513    Silvère Ganvoula  Silvère Ganvoula M'boussy   26    1.91   
514         514  Moritz Broschinski                        NaN   22    1.90   

    nationality   place_of_birth  price  max_price                 position  \
510       Japan      Komono, Mie   4.00        4.0    Attack - Right Winger   
511     Germany         Arnsberg   2.50        2.5  Attack - Centre-Forward   
512     Germany  Friedrichshafen   1.10        3.0  Attack - Centre-Forward   
513       Congo      Brazzaville   0.75        2.5  Attack - Centre-Forward   
514     Germany     Finsterwalde   0.50        0.5  Attack - Centre-Forward   

     shirt_nr   foot        club contract_expires joined_club  \
510        11  right  VfL Bochum       2024-06-30  2021-07-01   
511        33   left  VfL Bochum       2024-06-30  2022-07-01   
512         9  right  VfL Bochum       2024-06-30  2019-01-01   
513        35  right  VfL Bochum       2023-06-30  2019-07-05   
514        29  right  VfL Bochum       2026-06-30  2023-01-22   

              player_agent outfitter  
510  JEB ENTERTAINMENT ...    adidas  
511                   ROOF       NaN  
512                GoalSky      Puma  
513    BY Sport Consulting      Puma  
514          Teamgeist ...       NaN

Identificamos las columnas de la data

In [35]:
print(df.columns)

Index(['Unnamed: 0', 'name', 'full_name', 'age', 'height', 'nationality',
       'place_of_birth', 'price', 'max_price', 'position', 'shirt_nr', 'foot',
       'club', 'contract_expires', 'joined_club', 'player_agent', 'outfitter'],
      dtype='object')


In [36]:
df.describe()

Unnamed: 0         age      height       price   max_price    shirt_nr
count  515.000000  515.000000  515.000000  510.000000  510.000000  515.000000
mean   257.000000   25.683495    1.847612    8.482794   13.513627   19.798058
std    148.811962    4.700238    0.060996   14.653440   18.659222   12.414768
min      0.000000   17.000000    1.680000    0.025000    0.100000    1.000000
25%    128.500000   22.000000    1.800000    1.200000    2.500000    9.000000
50%    257.000000   25.000000    1.850000    3.500000    7.000000   20.000000
75%    385.500000   29.000000    1.890000    9.000000   16.750000   29.000000
max    514.000000   39.000000    2.000000  120.000000  150.000000   49.000000

In [37]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 515 entries, 0 to 514
Data columns (total 17 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   Unnamed: 0        515 non-null    int64  
 1   name              515 non-null    object 
 2   full_name         229 non-null    object 
 3   age               515 non-null    int64  
 4   height            515 non-null    float64
 5   nationality       515 non-null    object 
 6   place_of_birth    510 non-null    object 
 7   price             510 non-null    float64
 8   max_price         510 non-null    float64
 9   position          515 non-null    object 
 10  shirt_nr          515 non-null    int64  
 11  foot              506 non-null    object 
 12  club              515 non-null    object 
 13  contract_expires  481 non-null    object 
 14  joined_club       515 non-null    object 
 15  player_agent      488 non-null    object 
 16  outfitter         193 non-null    object 
dt

identificamos los tipo de dato de cada columna,
donde identificamos  float64(3), int64(3), object(11)

Ahora sumaremos la cantidad de nulos que existen en las columnas

In [38]:
df.isna().sum()

Unnamed: 0            0
name                  0
full_name           286
age                   0
height                0
nationality           0
place_of_birth        5
price                 5
max_price             5
position              0
shirt_nr              0
foot                  9
club                  0
contract_expires     34
joined_club           0
player_agent         27
outfitter           322
dtype: int64

Debido a que la cantidad de nulos que existen  y/o el enfoque analitico que buscamos se descarta las siguientes columnas. 
Datos que se desacartan:

 - full_name
 - Place of birth
 - Unnamed : 0




In [39]:
df.drop(['full_name', 'place_of_birth','Unnamed: 0','contract_expires','name','joined_club'], axis=1, inplace=True)

In [40]:
df.isna().sum()

name              0
age               0
height            0
nationality       0
price             5
max_price         5
position          0
shirt_nr          0
foot              9
club              0
joined_club       0
player_agent     27
outfitter       322
dtype: int64

Ahora identificaremos los valores unicos de la columna " outfitter "

In [41]:
df['outfitter'].unique()

array(['adidas', 'Puma', nan, 'Nike', 'New Balance', 'Uhlsport',
       'Under Armour', 'HashtagOne', 'Stanno', 'Mizuno'], dtype=object)

En el mundo del futbol las marcas realizan acuerdos con futbolistas reconocidos para dar valor a la marca por lo cual se sabe que es un grupo acotado y que el resto se considera como futbolista libre o Sin marca por lo cual los valores nulos se imputaran con el valor "free_sponsor"

In [42]:
df['outfitter'].fillna('free_sponsor', inplace=True)
df['outfitter'].unique()

array(['adidas', 'Puma', 'free_sponsor', 'Nike', 'New Balance',
       'Uhlsport', 'Under Armour', 'HashtagOne', 'Stanno', 'Mizuno'],
      dtype=object)

En el caso de los player_agent en los casos nulos se tomara la misma practica que con la marca, ya que en este entorno existen futbolistas que no son representados por empresas de representantes, caso como el de lionel messi que es representado por su padre. 
En este caso se le asignara el valor de "free_agent"

In [43]:
df['player_agent'].fillna('free_agent', inplace=True)
df['player_agent'].unique()



array(['PRO Profil GmbH', 'Relatives', 'BMS Sportconsulting ...',
       '11WINS', 'Rafaela Pimenta', 'Unique Sports Group',
       'Manuel García ...', 'ATG Sports', 'SEG', 'Gestifute', 'Carmenta',
       'BALLWERK Sports GmbH', 'Wasserman', 'free_agent',
       'Neubauer 13 GmbH', 'Team Raiola',
       'Agent is known - Player under 18', 'CAA Base Ltd', 'ROOF',
       'LIAN Sports Group', 'Kögl & Partner GmbH', 'ROGON',
       'SBE Management AG', 'AKA Global GmbH', 'SWSports',
       'Robert Schneider', 'Sports360 GmbH', 'HMH Sportmanagement',
       'Epic Sports', 'CAA Stellar', 'Thomas Meunier', 'Teamgeist ...',
       'Dirk Hebel', 'ISM Int. Soccer ...', 'Lorimanagement', 'Kick&Run',
       'Elite Project Group ...', 'management 360', 'HCM Sports ...',
       'Solbakken / Player ...', 'neblung ...', 'SOCCERNATION GROUP',
       'Marjan Sisic', 'BS Group - BS Law', 'Classico', 'TrueSports GmbH',
       'CN Sports', 'EMG Mundial', 'MR Sport Management ...',
       'Niagara Sports .

Al ser solo 9 nulos en en 'foot', utilizaremos la moda para imputar estos datos, en caso de desbalanceo realizaremos el control mas adelante

In [44]:
# Calcula la moda de la columna
moda_foot = df['foot'].mode()[0]

# Imputa los valores faltantes con la moda
df['foot'].fillna(moda_foot, inplace=True)

Para el caso de las columas que cuentan con valores numericos continuos, realizaremos:

Primero importamos la libreria de LinearRegression que nos probee sklearn 

In [15]:
from sklearn.linear_model import LinearRegression

Despues dividimos los datos entre valores conocidos (no nulos) y valores desconocidos (nulos)

In [16]:
data_known = df.dropna(subset=['price'])
data_unknown = df[df['price'].isnull()]

Ahora definimos las variables predictorias y la variable objetivo, comprobando la correlaciopn entre las variables

In [50]:
selected_columns = ['age', 'height']
correlation_with_target = df[selected_columns + ['price']].corr()['price']
print(correlation_with_target)

age      -0.103473
height   -0.075277
price     1.000000
Name: price, dtype: float64


Encontramos una correlacion negativa con las dos variable 

In [51]:
X_known = data_known[['age', 'height']]
y_known = data_known['price']
X_unknown = data_unknown[['age', 'height']]

Ahora realizamos el proceso de regresion 

In [68]:
# Crear un modelo de regresión lineal
regression_model = LinearRegression()

# Ajustar el modelo a los datos conocidos
regression_model.fit(X_known, y_known)

# Predecir los valores faltantes
predicted_values = regression_model.predict(X_unknown)

# Reemplazar los valores faltantes en el DataFrame original
df.loc[df['price'].isnull(), 'price'] = predicted_values

ValueError: Input y contains NaN.

Realizamos una copia para salverguardar el proceso

In [63]:
df1 = df.copy()

In [70]:
data_known = df1.dropna(subset=['max_price'])
data_unknown = df[df['max_price'].isnull()]

Ahora imputaremos 'Max price'

In [71]:
selected_columns = ['age', 'height', 'price']
correlation_with_target = df1[selected_columns + ['max_price']].corr()['max_price']
print(correlation_with_target)

age          0.093238
height      -0.089408
price        0.871134
max_price    1.000000
Name: max_price, dtype: float64


In [72]:
X_known = data_known[['age', 'height','price']]
y_known = data_known['max_price']
X_unknown = data_unknown[['age', 'height','price']]

In [75]:
# Crear un modelo de regresión lineal
regression_model = LinearRegression()

# Ajustar el modelo a los datos conocidos
regression_model.fit(X_known, y_known)

# Predecir los valores faltantes
predicted_values = regression_model.predict(X_unknown)

# Reemplazar los valores faltantes en el DataFrame original
df.loc[df['max_price'].isnull(), 'max_price'] = predicted_values

Validamos si aun existen nulos

In [76]:
df.isna().sum()

age             0
height          0
nationality     0
price           0
max_price       0
position        0
shirt_nr        0
foot            0
club            0
joined_club     0
player_agent    0
outfitter       0
dtype: int64

In [77]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 515 entries, 0 to 514
Data columns (total 12 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   age           515 non-null    int64  
 1   height        515 non-null    float64
 2   nationality   515 non-null    object 
 3   price         515 non-null    float64
 4   max_price     515 non-null    float64
 5   position      515 non-null    object 
 6   shirt_nr      515 non-null    int64  
 7   foot          515 non-null    object 
 8   club          515 non-null    object 
 9   joined_club   515 non-null    object 
 10  player_agent  515 non-null    object 
 11  outfitter     515 non-null    object 
dtypes: float64(3), int64(2), object(7)
memory usage: 48.4+ KB


Ahora cambiaremos el tipo de datos de las columnas de que poseen tipo object. 

In [78]:
#Nacionalidad
one_hot = pd.get_dummies(df['nationality'])
df = pd.concat([df, one_hot], axis=1)
print(df)

     age  height  nationality  price  max_price                 position  \
0     37    1.93      Germany   7.00       45.0               Goalkeeper   
1     34    1.83  Switzerland   5.00       13.0               Goalkeeper   
2     34    1.92      Germany   0.90        6.0               Goalkeeper   
3     20    1.91      Germany   0.30        0.3               Goalkeeper   
4     23    1.89  Netherlands  75.00       75.0   Defender - Centre-Back   
..   ...     ...          ...    ...        ...                      ...   
510   28    1.73        Japan   4.00        4.0    Attack - Right Winger   
511   30    1.95      Germany   2.50        2.5  Attack - Centre-Forward   
512   31    1.79      Germany   1.10        3.0  Attack - Centre-Forward   
513   26    1.91        Congo   0.75        2.5  Attack - Centre-Forward   
514   22    1.90      Germany   0.50        0.5  Attack - Centre-Forward   

     shirt_nr   foot           club joined_club  ... Turkey Turkey  Germany  \
0       

In [79]:
#Position 
one_hot = pd.get_dummies(df['position'])
df = pd.concat([df, one_hot], axis=1)
print(df)

     age  height  nationality  price  max_price                 position  \
0     37    1.93      Germany   7.00       45.0               Goalkeeper   
1     34    1.83  Switzerland   5.00       13.0               Goalkeeper   
2     34    1.92      Germany   0.90        6.0               Goalkeeper   
3     20    1.91      Germany   0.30        0.3               Goalkeeper   
4     23    1.89  Netherlands  75.00       75.0   Defender - Centre-Back   
..   ...     ...          ...    ...        ...                      ...   
510   28    1.73        Japan   4.00        4.0    Attack - Right Winger   
511   30    1.95      Germany   2.50        2.5  Attack - Centre-Forward   
512   31    1.79      Germany   1.10        3.0  Attack - Centre-Forward   
513   26    1.91        Congo   0.75        2.5  Attack - Centre-Forward   
514   22    1.90      Germany   0.50        0.5  Attack - Centre-Forward   

     shirt_nr   foot           club joined_club  ... Attack - Second Striker  \
0      

In [80]:
#foot 
one_hot = pd.get_dummies(df['foot'])
df = pd.concat([df, one_hot], axis=1)
print(df)

     age  height  nationality  price  max_price                 position  \
0     37    1.93      Germany   7.00       45.0               Goalkeeper   
1     34    1.83  Switzerland   5.00       13.0               Goalkeeper   
2     34    1.92      Germany   0.90        6.0               Goalkeeper   
3     20    1.91      Germany   0.30        0.3               Goalkeeper   
4     23    1.89  Netherlands  75.00       75.0   Defender - Centre-Back   
..   ...     ...          ...    ...        ...                      ...   
510   28    1.73        Japan   4.00        4.0    Attack - Right Winger   
511   30    1.95      Germany   2.50        2.5  Attack - Centre-Forward   
512   31    1.79      Germany   1.10        3.0  Attack - Centre-Forward   
513   26    1.91        Congo   0.75        2.5  Attack - Centre-Forward   
514   22    1.90      Germany   0.50        0.5  Attack - Centre-Forward   

     shirt_nr   foot           club joined_club  ... Defender - Right-Back  \
0        

Evaluaremos la cantidad de filas y columnas que hemos generado. 

In [81]:
df.shape

(515, 175)

Con el cambio de tipo de datos hemos alcanzdo un total de 175 columnas. 